### PR TITLE
workflows: Add step which extracts authors from PDFs

### PR DESCRIPTION
**Pull request by pazembrz**
This requires a new env variable `GROBID_URL` which provides url to grobid (I will add it to kubernetes repo next week)

### CODE REVIEW

File: inspirehep/modules/workflows/tasks/actions.py [hunk 1]
```diff
@@ -738,3 +739,33 @@ def replace_collection_to_hidden(obj, eng):
...
+    acquisition_source = get_value(obj.data, 'acquisition_source.source', '').lower()
+    if acquisition_source in ['arxiv', 'pos'] and not get_mark(obj, "authors_xml", False):
+        return True
+    return False
+
+
+@with_debug_logging
+def extract_authors_from_pdf(obj, eng):
+    with get_document_in_workflow(obj) as tmp_document:
+        if tmp_document:
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply inspirehep/modules/workflows/tasks/actions.py:756]: :+1:

File: inspirehep/modules/workflows/tasks/actions.py [hunk 2]
```diff
@@ -738,3 +739,33 @@ def replace_collection_to_hidden(obj, eng):
...
+                document = document_file.read()
+            data = {'input': document}
+            data.update({"includeRawAffiliations": "1", "consolidateHeader": "1"})
+            grobid_url = current_app.config["GROBID_URL"]
+            api_path = "api/processHeaderDocument"
+            response = requests.post(urljoin(grobid_url, api_path), files=data)
+            response.raise_for_status()
+            authors_and_affiliations = GrobidAuthors(response.text)
+
+            if len(authors_and_affiliations) > 0 and len(authors_and_affiliations) == len(obj.data.get('authors')):
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply inspirehep/modules/workflows/tasks/actions.py:767]: Logging is OK but reverted IF statement is much less readable IMHO. I will leave the original one.

File: inspirehep/modules/workflows/utils/__init__.py [hunk 3]
```diff
@@ -530,3 +530,16 @@ def post_record_to_hep(pid_type, data=None):
...
     return response.json()
+
+
+def set_mark(obj, key, value):
+    obj.extra_data[key] = value
+    return {key: value}
+
+
+def check_mark(obj, key):
+    return key in obj.extra_data and obj.extra_data[key]
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply inspirehep/modules/workflows/utils/__init__.py:541]: :+1:

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+            yield {
+                'author': self._builder.make_author(
+                    full_name=author.fullname,
+                    raw_affiliations=author.raw_affiliations,
+                    emails=author.emails,
+                ),
+                'parsed_affiliations': author.processed_affiliations
+            }
+
+    def getall(self):
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply inspirehep/modules/workflows/utils/grobid_authors_parser.py:57]: :+1:

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        self._parsed_authors = self._xml.xpath("//author")
+        self._builder = None
+
+    def __getitem__(self, item):
+        return GrobidAuthor(self._parsed_authors[item])
+
+    def __len__(self):
+        return len(self._parsed_authors)
+
+    def get(self):
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply inspirehep/modules/workflows/utils/grobid_authors_parser.py:44]: This sounds odd. Maybe `parse_one`?

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        data = self._extract(source, path, type, text=True).getall()
+        if clean:
+            data = [text.strip() for text in data]
+        return join_char.join(data)
+
+    def _extract_strings_list(self, source, path, type=None, clean=True):
+        data = self._extract(source, path, type, text=True).getall()
+        if clean:
+            return [text.strip() for text in data]
+        return data
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply inspirehep/modules/workflows/utils/grobid_authors_parser.py:87]: IMHO it's easier to not care about extracting/joining in so many places. And code is much more readable as it's like 20 times smaller.

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    def names(self):
+        return self._extract_string(self._author, "persName/forename")
+
+    @property
+    def lastname(self):
+        return self._extract_string(self._author, "persName/surname")
+
+    @property
+    def fullname(self):
+        return u" ".join([self.names, self.lastname])
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply inspirehep/modules/workflows/utils/grobid_authors_parser.py]: :+1:

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+                ),
+                'parsed_affiliations': author.processed_affiliations
+            }
+
+    def getall(self):
+        """Returns all authors at once as a list"""
+        parsed_authors = []
+        for author_data in self.get():
+            parsed_authors.append(author_data)
+        return parsed_authors
```
[the remark this answers is not in the document]
Comment (pazembrz) [reply inspirehep/modules/workflows/utils/grobid_authors_parser.py]: :+1:

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    def parse_all(self):
+        """Returns all authors at once as a list"""
+        return list(self.parse_one())
+
+
+class GrobidAuthor(object):
+    def __init__(self, author_selector):
+        self._author = author_selector
+
+    def _extract(self, source, path, type=None, text=False):
```
Comment (michamos) [inline inspirehep/modules/workflows/utils/grobid_authors_parser.py:66]: not using `self`, so `@staticmethod`

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        if type:
+            path += "[@type='{type}']".format(type=type)
+        if text:
+            path += "/text()"
+            return source.xpath(path)
+        return source.xpath(path)
+
+    def _extract_string(self, source, path, type=None, join_char=u' ', clean=True):
```
Comment (michamos) [inline inspirehep/modules/workflows/utils/grobid_authors_parser.py:74]: not using the instance data (`self` is needed to access `_extract` only), so better make this a `@classmethod` instead). Also, isn't this just `join_char.join(_extract_strings_list(...))`?

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+        data = self._extract(source, path, type, text=True).getall()
+        if clean:
+            data = [text.strip() for text in data]
+        return join_char.join(data)
+
+    def _extract_strings_list(self, source, path, type=None, clean=True):
```
Comment (michamos) [inline inspirehep/modules/workflows/utils/grobid_authors_parser.py:80]: same

**Review by michamos — APPROVED** [review]
some small comments, looks good otherwise :+1:

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py (hunk 9 above)
Comment (pazembrz) [reply inspirehep/modules/workflows/utils/grobid_authors_parser.py:66]: :+1:

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py (hunk 11 above)
Comment (pazembrz) [reply inspirehep/modules/workflows/utils/grobid_authors_parser.py:80]: :+1:

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py (hunk 9 above)
Comment (pazembrz) [reply inspirehep/modules/workflows/utils/grobid_authors_parser.py:66]: :+1:

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py (hunk 10 above)
Comment (pazembrz) [reply inspirehep/modules/workflows/utils/grobid_authors_parser.py:74]: It's more like: ` join_char.join([x.strip() for x in _extract_strings_list(...)])` It's cleaner to call this method 6 times in a row (when building address) than calling original method in a join call with strip in a loop :P .

File: inspirehep/modules/workflows/utils/grobid_authors_parser.py (hunk 11 above)
Comment (pazembrz) [reply inspirehep/modules/workflows/utils/grobid_authors_parser.py:80]: :+1:

**Review by michamos — APPROVED** [review]

**Review by michamos — APPROVED** [review]

**Review by michamos — APPROVED** [review]

**Pull request merged**